src/Reserve.sol

```diff
@@ -4,30 +4,115 @@ pragma solidity ^0.8.7;
+    /// @param amt The deposited amount.
+    function deposit(
         IERC20 token,
-        address to,
+        address from,
         uint256 amt
     ) external;
 
-    function deposit(
+    /// @notice Withdraws funds from the reserve.
+    /// The reserve will transfer `amt` tokens to the `to` address.
+    /// Only funds previously deposited can be withdrawn.
+    /// @param token The used token.
+    /// @param to The address to which funds are withdrawn.
+    /// @param amt The withdrawn amount.
+    function withdraw(
         IERC20 token,
-        address from,
+        address to,
         uint256 amt
     ) external;
 }
 
+/// @notice The reserve plugin interface required by the reserve.
+interface IReservePlugin {
+    /// @notice Called by the reserve immediately after transferring funds for deposition.
+    /// @param token The used token.
+    /// @param amt The amount which has been transferred for deposition.
+    function afterDeposition(IERC20 token, uint256 amt) external;
+
+    /// @notice Called by the reserve right before transferring funds for withdrawal.
+    /// The reserve will `transferFrom` the tokens from the plugin address.
+    /// The reserve can always withdraw everything that has been ever deposited, but never more.
+    /// @param token The used token.
+    /// @param amt The amount which will be transferred.
+    function beforeWithdrawal(IERC20 token, uint256 amt) external;
+}
+
+/// @notice The ERC-20 tokens reserve contract.
+/// The registered users can deposit and withdraw funds.
+/// The reserve by default doesn't do anything with the tokens,
+/// but for each ERC-20 address a plugin can be registered for tokens storage.
 contract Reserve is IReserve, Ownable {
+    /// @notice The dummy plugin address meaning that no plugin is being used.
+    IReservePlugin public constant NO_PLUGIN = IReservePlugin(address(0));
+
+    /// @notice A set of addresses considered users.
+    /// The value is `true` if an address is a user, `false` otherwise.
     mapping(address => bool) public isUser;
-    mapping(IERC20 => uint256) public withdrawable;
+    /// @notice How many tokens are deposited for each token address.
+    mapping(IERC20 => uint256) public deposited;
+    /// @notice The reserved plugins for each token address.
+    mapping(IERC20 => IReservePlugin) public plugins;
 
-    event Withdrawn(address user, IERC20 indexed token, address indexed to, uint256 amt);
+    /// @notice Emitted when a plugin is set.
+    /// @param owner The address which called the function.
+    /// @param token The token for which plugin has been set.
+    /// @param oldPlugin The old plugin address. `NO_PLUGIN` if no plugin was being used.
+    /// @param newPlugin The new plugin address. `NO_PLUGIN` if no plugin will be used.
+    event PluginSet(
+        address owner,
+        IERC20 indexed token,
+        IReservePlugin indexed oldPlugin,
+        IReservePlugin indexed newPlugin
```

Comment:
I think it would the nice to add the `amount` we took from old plugin and moved into the new plugin to the event.

Reply:
Added